hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/audit/S3AAuditLogMergerAndParser.java

```diff
@@ -0,0 +1,546 @@
+
+    dataFileWriter.create(avroDataRecord.getSchema(),
+        new File("AvroData.avro"));
+
+    ArrayList<String> longValues = new ArrayList<>(
+        Arrays.asList("turnaroundtime", "bytessent", "objectsize",
+            "totaltime"));
+    int count = 0;
+
+    // Write avro data into a file in bucket destination path
+    Path avroFile = new Path(s3DestPath, "AvroData.avro");
+
+    try (FSDataOutputStream fsDataOutputStream = s3AFileSystem.create(
+        avroFile)) {
+      // Insert data according to schema
+      for (Map<String, String> auditLogMap : auditLogList) {
+        for (Map.Entry<String, String> entry : auditLogMap.entrySet()) {
+          String key = entry.getKey();
+          String value = entry.getValue().trim();
+
+          // if value == '-' and key is not in arraylist then put '-' or else '-1'
+          // if key is in arraylist of long values then parse the long value
+          // while parsing do it in try-catch block,
+          // in catch block need to log exception and set value as '-1'
+          try {
+            if (longValues.contains(key)) {
+              if (value.equals("-")) {
+                avroDataRecord.put(key, null);
+              } else {
+                avroDataRecord.put(key, Long.parseLong(value));
+              }
+            } else {
+              avroDataRecord.put(key, value);
+            }
+          } catch (Exception e) {
+            avroDataRecord.put(key, null);
+          }
+        }
+        avroDataRecord.put("referrerMap", referrerHeaderList.get(count));
+        dataFileWriter.append(avroDataRecord);
+        byte[] byteArray = avroDataRecord.toString().getBytes("UTF-8");
+        fsDataOutputStream.write(byteArray);
+        count += 1;
+      }
+    }
+    dataFileWriter.close();
+    LOG.info("Successfully generated avro data");
+  }
+
+  /**
+   * Merge and parse all the audit log files and convert data into avro file.
+   *
+   * @param s3AFileSystem filesystem
+   * @param s3LogsPath    source path of logs
+   * @param s3DestPath    destination path of merged log file
+   * @return true
+   * @throws IOException on any failure
+   */
+  public boolean mergeAndParseAuditLogFiles(S3AFileSystem s3AFileSystem,
+      Path s3LogsPath,
+      Path s3DestPath) throws IOException {
+
+    RemoteIterator<LocatedFileStatus> listOfS3LogFiles =
+        s3AFileSystem.listFiles(s3LogsPath, true);
+
+    Path destFile = new Path(s3DestPath, "AuditLogFile");
+
+    List<HashMap<String, String>> entireAuditLogList = new ArrayList<>();
+    List<HashMap<String, String>> referrerHeaderList = new ArrayList<>();
+    List<HashMap<String, String>> auditLogList = new ArrayList<>();
+
+    try (FSDataOutputStream fsDataOutputStream = s3AFileSystem.create(
+        destFile)) {
+      // Iterating over the list of files to merge and parse
+      while (listOfS3LogFiles.hasNext()) {
+        Path logFile = listOfS3LogFiles.next().getPath();
```

Comment:
listOfS3LogFiles.next() is a file status, so no need to ask for it again on the next line
```java
FileStatus fileStatus = listOfS3LogFiles.next()
```

Reply:
modified